File: tests/fixtures/ecosystems/ruby_comprehensive/inheritance.rb
# Inheritance patterns for parser testing.

module Comprehensive
  class Animal
    attr_reader :name

    def initialize(name)
      @name = name
    end

    def speak
      raise NotImplementedError, "Subclass must implement speak"
    end

    def describe
      "#{@name} says #{speak}"
    end
  end

  class Dog < Animal
    def speak
      "Woof!"
    end

    def fetch(item)
      "#{@name} fetched #{item}"
    end
  end

  class Cat < Animal
    def speak
      "Meow!"
    end
  end

  module Loggable
    def log(message)
      puts "[#{self.class.name}] #{message}"
    end
  end

  module Serializable
    def to_hash
      instance_variables.each_with_object({}) do |var, hash|
        hash[var.to_s.delete("@")] = instance_variable_get(var)
      end
    end
  end

  class ServiceDog < Dog
    include Loggable
    include Serializable

    attr_reader :job

    def initialize(name, job)
      super(name)
      @job = job
    end

    def work
      log("Working as #{@job}")
      "#{@name} is working"
    end
  end
end

File: tests/fixtures/ecosystems/python_comprehensive/inheritance.py
"""Inheritance patterns for parser testing."""

from abc import ABC, abstractmethod


class Animal(ABC):
    """Abstract base class for animals."""

    def __init__(self, name: str):
        self.name = name

    @abstractmethod
    def speak(self) -> str:
        ...

    def describe(self) -> str:
        return f"{self.name} says {self.speak()}"


class Dog(Animal):
    """Dog inherits from Animal."""

    def speak(self) -> str:
        return "Woof!"

    def fetch(self, item: str) -> str:
        return f"{self.name} fetched {item}"


class Cat(Animal):
    def speak(self) -> str:
        return "Meow!"


class LogMixin:
    """Mixin for logging."""

    def log(self, message: str) -> None:
        print(f"[{self.__class__.__name__}] {message}")


class SerializeMixin:
    """Mixin for serialization."""

    def to_dict(self) -> dict:
        return {"class": self.__class__.__name__}


class ServiceDog(Dog, LogMixin, SerializeMixin):
    """Multiple inheritance: Dog + LogMixin + SerializeMixin."""

    def __init__(self, name: str, job: str):
        super().__init__(name)
        self.job = job

    def work(self) -> str:
        self.log(f"Working as {self.job}")
        return f"{self.name} is working"


class GuideDog(ServiceDog):
    """Deep inheritance chain."""

    def guide(self, destination: str) -> str:
        return f"Guiding to {destination}"
